src/libmeasurement_kit/ooni/facebook_messenger.cpp

```diff
@@ -0,0 +1,220 @@
+// Part of measurement-kit <https://measurement-kit.github.io/>.
+// Measurement-kit is free software. See AUTHORS and LICENSE for more
+// information on the copying conditions.
+
+#include "private/common/fcompose.hpp"
+#include "private/common/utils.hpp"
+#include "private/ooni/utils.hpp"
+#include "private/ooni/constants.hpp"
+#include <measurement_kit/ooni.hpp>
+
+namespace mk {
+namespace ooni {
+
+using namespace mk::report;
+
+static void dns_many(Error error,
+                     Var<Entry> entry,
+                     std::map<std::string, std::string> fb_service_hostnames,
+                     Var<Reactor> reactor,
+                     Var<Logger> logger,
+                     Callback<Error,
+                              Var<Entry>,
+                              std::map<std::string, std::vector<std::string>>,
+                              Var<Reactor>,
+                              Var<Logger>> cb
+                    ) {
+    Var<std::map<std::string, std::vector<std::string>>>
+        fb_service_ips(new std::map<std::string, std::vector<std::string>>(
+            {{"stun", {}},
+             {"b_api", {}},
+             {"b_graph", {}},
+             {"edge", {}},
+             {"external_cdn", {}},
+             {"scontent_cdn", {}},
+             {"star", {}}}
+        ));
+
+    if (error) {
+        cb(error, entry, *fb_service_ips, reactor, logger);
+        return;
+    }
+
+    // if we find any inconsistent DNS later, switch this to true
+    (*entry)["facebook_dns_blocking"] = false;
+
+    int names_count = fb_service_hostnames.size();
+    Var<int> names_tested(new int(0));
+
+    if (names_count == *names_tested) {
+        cb(NoError(), entry, *fb_service_ips, reactor, logger);
+        return;
+    }
+
+    auto dns_cb = [=](std::string service, std::string hostname) {
+        return [=](Error err, Var<dns::Message> message) {
+            if (!!err) {
+                logger->info("fb_messenger: dns error for %s, %s",
+                    service.c_str(), hostname.c_str());
+            } else {
+                for (auto answer : message->answers) {
+                    if ((answer.ipv4 != "") || (answer.hostname != "")) {
+                        //XXX
+                        //std::string asn_p = options.get("geoip_asn_path", std::string{});
+                        std::string asn_p = "GeoIPASNum.dat";
+                        GeoipDatabase geoip(asn_p);
+                        ErrorOr<std::string> asn = geoip.resolve_asn(answer.ipv4);
```

Comment:
Here you should:

1. fetch the geoip file name from options (you should pass them as argument to the function running the test and then pass them around)

2. use the thread-local geoip cache

You can see how this is done in the implementation of web connectivity.

Reply:
done